lib/fog/compute/xen_server/real.rb

```diff
@@ -14,7 +14,15 @@ def initialize(options={})
           @port        = options[:xenserver_port] || 80
           @verify_mode = options[:xenserver_verify_mode] || OpenSSL::SSL::VERIFY_PEER
           @connection  = Fog::XenServer::Connection.new(@host, @port, @use_ssl, @verify_mode, @timeout)
+
           @connection.authenticate(@username, @password)
+
+          host_master = @connection.find_pool_master
+          if host_master
+            @host = host_master
+            @connection = Fog::XenServer::Connection.new(@host, @timeout)
+            @connection.authenticate(@username, @password)
+          end
```

Comment:
I believe it would be best to have a method to return if the connection was made on a slave node in the `Connection` object and another methods to return the master like:

```ruby
def slave?
  response = @factory.call("host.get_all_records", @credentials)
  response["Status"] == "Failure"
end

def master
  response = @factory.call("host.get_all_records", @credentials)
  return if response["ErrorDescription"].nil?
  return unless response["ErrorDescription"][0] == "HOST_IS_SLAVE"
  response["ErrorDescription"][1]
end
```

Then we could define the `Real` class as:

```ruby
class Real
  def initialize(options = {})
    .
    .
    .
    .
    @connection  = Fog::XenServer::Connection.new(@host, @port, @use_ssl, @verify_mode, @timeout)

    @connection.authenticate(@username, @password)

    if @connection.slave?
      @connection  = Fog::XenServer::Connection.new(@connection.master, @port, @use_ssl, @verify_mode, @timeout)
      @connection.authenticate(@username, @password)
    end
  end
end
```

WDYT

Reply:
You're right :+1: